[Woo POS][Historical Orders] Order Details (Wireframe UI)

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

WOOMOB-1138

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Implement order details UI according to the wireframes. Include a header section with Order ID, Date & Time, Total, Status, Payment Method, and Customer Email. Display the products section and totals section. No actions are included in this task.

- Expanded `POSOrderItem` with additional required fields
- Developed `PointOfSaleOrderDetailsView` with expanded header, products, and totals sections
- Created `PointOfSaleOrderDetailsViewHelper` to consolidate formatting and calculations logic
- Improved state management of selected orders
- Added simple loading and empty views, they will be finalized in another task

One thing that I didn't include are product images, since they require additional logic I haven't anticipated. I'l added it into a new PR.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Open POS -> Menu -> Orders
2. Confirm the loading state of a list and order details
3. Confirm that order details automatically appear after finishing loading
4. Select orders of different states - completed, failed, refunded, or partially refunded
5. Confirm correct status, products, product attributes, total, paid, and refund (if applicable) calculations are shown

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

iPad Air 18.5 Simulator

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### Loading and navigating orders


https://github.com/user-attachments/assets/f7f2657a-15fd-4d40-99c7-5a95c9b8faad


### Making a new order & reopening


https://github.com/user-attachments/assets/aff28ce9-aa03-46eb-bd52-d896c3e61713


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Expand POSOrder and POSOrderItem for details view needs

staskus committed:
Allow setting bottom content on POSPageHeaderView

staskus committed:
Create PointOfSaleOrderDetailsViewHelper to handle calculations and formatting

staskus committed:
Add PointOfSaleOrderDetailsViewHelperTests

staskus committed:
Improve PointOfSaleOrderDetailsView to wireframe designs

staskus committed:
Manage selected order within orders controller

staskus committed:
Add a simple loading view for Order Details

staskus committed:
Move PointOfSaleOrderBadgeView into a separate view to reuse

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/254">23.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Merge branch 'trunk' into woomob-1138-woo-poshistorical-orders-order-details-wireframe-ui

staskus committed:
Remove unused code

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSPageHeaderView.swift:
We got new warnings here, it seems to detect that we had a trailing closure for `trailingContent` previously, and now is updated with a new `bottomContent ` trailing closure, so it wants to be declared explicitly: 

```
Backward matching of the unlabeled trailing closure is deprecated; label the argument with 'bottomContent' to suppress this warning
```

iamgabrielma commented on WooCommerce/Classes/POS/ViewHelpers/PointOfSaleOrderDetailsViewHelper.swift:
```diff
@@ -0,0 +1,106 @@
+
+    /// Calculates the products subtotal by summing line item subtotals
+    /// Follows the same pattern as OrderPaymentDetailsViewModel.subtotal
+    func productsSubtotal(for order: POSOrder) -> String {
```
Not necessarily as part of this PR (or project), but to avoid repetition on those across features, I wonder if would be worth to move these calculations as functions over POSOrder, rather than keep them as free functions. So rather than `helper.productsSubtotal(for: posOrder)` we would do `posOrder.productsSubtotal()` 🤔 WDYT? I'm happy to open some experimentation branch with this if makes sense.

staskus replied:
Good idea, I'll see what could be done now 👍

staskus replied:
I ended up following this advice. I found `Order+CurrencyFormattedValues`, expanded and improved it a bit (fixing some formatting inconsistencies in Woo App order details as well), and then created `POSOrdersMapper` to set formatted values straight on `POSOrder`, allowing views to use these values. It simplified the code, reused existing logic, and avoided using CurrencyFormatter (and ServiceLocator) in views.

iamgabrielma commented on WooCommerce/Classes/POS/ViewHelpers/PointOfSaleOrderDetailsViewHelper.swift:
```diff
@@ -0,0 +1,106 @@
+struct PointOfSaleOrderDetailsViewHelper {
+    private let currencyFormatter: CurrencyFormatter
+
+    init(currencySettings: CurrencySettings = ServiceLocator.currencySettings) {
```
Not sure how much this is worth the efford, but should we provide a default value for currencySettings so deep in the view or move it higher? Thinking in the context of getting rid of direct references to ServiceLocator for demodularization

Since `PointOfSaleOrderDetailsView()` owns `PointOfSaleOrderDetailsViewHelper()` perhaps is worth to inject it a bit higher. The DI chain is a bit annoying here, so perhaps could sit in `PointOfSaleOrderListModel` or `PointOfSaleOrderListController` 🤔

staskus replied:
Yes, I agree. I dealt with it during modularization work. What I did was inject the currency settings protocol into the environment so it could be accessed within views.  It may be worth bringing some of the modularization helper code to trunk, not to introduce more dependencies...

iamgabrielma commented on WooCommerce/Classes/POS/ViewHelpers/PointOfSaleOrderDetailsViewHelper.swift:
```diff
@@ -0,0 +1,106 @@
+    /// Formats discount total
+    /// Follows TotalsView pattern using discountTotal directly
+    func formattedDiscountTotal(for order: POSOrder) -> String? {
+        guard let discountTotal = Double(order.discountTotal), discountTotal != 0 else {
```
Just double checking since the original implementation checks for `>0`. Is `!=0` what we want here? That would allow for negative discounts depending on how is used, which would be summed to the amount `(amount -  (-discount))`

iamgabrielma commented on WooCommerce/Classes/POS/ViewHelpers/PointOfSaleOrderDetailsViewHelper.swift:
```diff
@@ -0,0 +1,106 @@
+
+    /// Formats tax total
+    func formattedTaxTotal(for order: POSOrder) -> String? {
+        guard let taxTotal = Double(order.totalTax), taxTotal != 0 else {
```
Same comment as above regarding `>0` vs `!=0`

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/ViewHelpers/PointOfSaleOrderDetailsViewHelperTests.swift:
```diff
@@ -0,0 +1,471 @@
+
+    // MARK: - Products Subtotal Tests
+
+    @Test func productsSubtotal_emptyLineItems_returnZero() async throws {
```
I'd add the "formatted" bit to the test name: 

```suggestion
    @Test func productsSubtotal_when_empty_line_items_then_returns_formatted_zero_subtotal() async throws {
```

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/ViewHelpers/PointOfSaleOrderDetailsViewHelperTests.swift:
In general, if possible, I'd add when/then to all the test names in the file as well as separate the words that are not defined as variables within the test, ie: singleItem -> `single_item`

iamgabrielma commented on WooCommerce/WooCommerceTests/POS/ViewHelpers/PointOfSaleOrderDetailsViewHelperTests.swift:
```diff
@@ -0,0 +1,471 @@
+        #expect(result == false)
+    }
+
+    @Test func shouldShowDiscount_positiveDiscount_returnsTrue() async throws {
```
Let's add a test for negative discounts as well, my understanding is that the negative discount should not be supported, as we detract it from the order amount so its always absolute, but if a negative value happens to go through we may want to handle it here as well.

iamgabrielma commented on WooCommerce/Classes/POS/Controllers/PointOfSaleOrderListController.swift:
```diff
@@ -118,4 +121,9 @@ protocol PointOfSaleOrderListControllerProtocol {
         ordersViewState = .loading(cachedOrders)
     }
+
+    @MainActor
```
This one does not require MainActor, if we want to run it in the main actor then we need to make the function and conformance async 
```suggestion
```

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSPageHeaderView.swift:
```diff
@@ -38,10 +38,11 @@ struct POSPageHeaderItem: Identifiable {
 /// A header view for POS pages.
 /// Design ref: 1qcjzXitBHU7xPnpCOWnNM-fi-450_24951
-struct POSPageHeaderView<TrailingContent: View>: View {
+struct POSPageHeaderView<TrailingContent: View, BottomContent: View>: View {
```
💯

iamgabrielma reviewed (approved):
Works well! Looking good 🚢

staskus committed:
Fix warning: Backward matching of the unlabeled trailing closure is deprecated

staskus committed:
Rename OrdersViewState to POSOrdersViewState

staskus committed:
Remove PointOfSaleOrderDetailsViewHelper

staskus committed:
Expand Order+CurrencyFormattedValues with ability to inject currency formatter

Injecting currency formatter allows to correctly format totals in the view layer

staskus committed:
Create POSOrderMapper to set formatted and calculated totals on POSOrder

staskus committed:
Use formatted values from order in list and details view

staskus committed:
Update PreviewHelpers.swift

staskus committed:
Update tests

staskus committed:
Remove unused code

staskus committed:
Create POSOrderMapperTests

staskus committed:
Remove errors when fetching images

staskus committed:
Update PointOfSaleOrderDetailsView.swift

staskus committed:
Remove unused code
